src/coreclr/jit/morph.cpp

```diff
@@ -11683,19 +11682,15 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
 
             bool foldAndReturnTemp = false;
             temp                   = nullptr;
-            ival1                  = 0;
 
             // Don't remove a volatile GT_IND, even if the address points to a local variable.
-            if ((tree->gtFlags & GTF_IND_VOLATILE) == 0)
+            //
+            if (!tree->AsIndir()->IsVolatile())
             {
                 /* Try to Fold *(&X) into X */
                 if (op1->gtOper == GT_ADDR)
                 {
-                    // Can not remove a GT_ADDR if it is currently a CSE candidate.
-                    if (gtIsActiveCSE_Candidate(op1))
-                    {
-                        break;
-                    }
+                    assert(!gtIsActiveCSE_Candidate(op1)); // GT_ADDRs cannot be CSE candidates.
```

Comment:
Hmm, why not?

Reply:
Immediate reason is that `optIsCSEcandidate` always returns `false` for them.

Reasons why wouldn't want / be able to CSE them:
1) CSEing `ADDR`s that point to tracked locals would break things.
2) CSEing `ADDR`s that point to address-exposed locals is likely not very useful because they're cheap to rematerialize (with the exception of non-XARCH platforms with huge frames, I suppose).
3) CSEing `ADDR(HWI/SIMD)` would break things.
4) All other `ADDR`s (`ADDR(IND/BLK/OBJ)`) are pretty much transient (get created and then immediately morphed away).

Reply:
I don't really see a good reason to bake an implicit assumption that we are never going to be able to CSE `GT_ADDR` nodes, e.g. in a case like: https://godbolt.org/z/fWEhhGaoK

Reply:
I don't think it is unreasonable to CSE local addresses, but I do not see us doing that with `ADDR`s, because they will increasingly only get used for tracked locals. Note: your example would have `LCL_VAR_ADDR` nodes.

Reply:
> but I do not see us doing that with ADDRs, because they will increasingly only get used for tracked locals

Well, ideally that should not matter. The same example with a promoted struct shows the same codegen difference between Clang/GCC and RyuJIT.

Reply:
> Well, ideally that should not matter

Do you mean not matter as in that we should be able to CSE even addresses of tracked locals? That seems like a hard problem.

Regardless, I will put the CSE check back in.

Reply:
> Do you mean not matter as in that we should be able to CSE even addresses of tracked locals? That seems like a hard problem.

I agree, the bookkeeping required for this would surely make it infeasible/not worth it. I'd just hope this wasn't a global invariant (but given that you are way more familiar with the story around `GT_ADDR` and the other addressing nodes, maybe this hope is unfounded :) ).